scripts/scraper-ng/rules/html-require-recipe-ingredients/ingredient-handlers/data-static-methods.js

```diff
@@ -0,0 +1,27 @@
+const { select } = require("hast-util-select");
+
+const utils = require("./utils.js");
+const checkLinkList = require("./link-list-handler.js");
+
+/**
+ * Handler for the `data.specifications` ingredient.
+ */
+function handleDataStaticMethods(tree, file, context) {
+  const id = "Static_methods";
+  const body = select(`body`, tree);
+  const heading = select(`h2#${id}`, body);
+
+  // This is an optional ingredient, so if there's no `h2`,
+  // assume that the page intends to omit it.
+  if (heading === null) {
+    return;
+  }
+
+  const section = utils.sliceSection(heading, body);
+  // the first element is always the `h2`, which we are not interested in
+  const children = section.children.slice(1);
+  const logError = utils.logError(body, file, context);
```

Comment:
> there's a more concrete problem

Oh, OK. I hadn't realised that this was what this argument was for. I guess that's evidence that error reporting here is quite complex. Of course this would be easy to fix by making the node an argument to the logging function (like `text` and `id` are now) rather than an argument to the factory: that is, it's a thing you know at the point you log the error.

But on the general points: fair enough if this seems too complex. But I think there is a problem in this code at the moment. First, as I understand it, lots of places are going to have to write code like:

```js
const message = file.message(
  `Expected h2#${id}`,
  body,
  `${context.source}:${context.recipeName}/${context.ingredient}/expected-heading`
);
message.fatal = true;
utils.logMissingIngredient(file, context);
```

So that's a lot of duplication, and a lot of scope for error messages to be inconsistently formatted from one place to another.

Second, unless I misunderstand this, anywhere that's doing this is going to need the `file` and `context` objects. So at the moment we're passing these things around. And if we want to factor common bits of linting into separate modules, we have to pass them around there as well. It's just a lot of state to ship around.

Reply:
Yes, I agree that the places where we call `file.message()` are high risk for inconsistency and that we're currently passing around a lot of state. You might be on the right track of creating a logger at the top of a handler and passing it around, but maybe we can work on the API a bit. Something like this?

```javascript
const logger = new IngredientLogger(file, context);
logger.fail() // fails the whole ingredient, using the whole file as a position, like utils.logMissingIngredient
logger.fail(someNode) // fails the whole ingredient, using a specific node as a position, such as an h2
logger.expected(someNode, "compat-macro") // logs an "expected-compat-macro" message against someNode, then calls `logger.fail` too
```

Or a more half-baked idea I've had would be to make all the logging happen at the top of the plugin and have ingredient handlers throw an exception type. Then you could write some module-level `Error` that the plugin would turn into a vfile message.

Reply:
How do you want to proceed with this?

1. refactor logging as part of this PR  
2. separate PRs:
    * first, continue with this PR using the current logging
    * then refactor it in a follow-up
3. separate PRs:
    * first, refactor logging
    * then continue with this PR

The problem with (1): bigger PR, less focused, more disparate discussions.
The problem with (2): this PR will get ugly without some attempt to tame the logging code.
The problem with (3): this PR is pushed further back.

I think (3) is probably best.

Reply:
Yeah, let's refactor logging, then go from there. I'll plan to review as quickly as possible to minimize the delay.

Reply:
I've filed https://github.com/mdn/stumptown-content/issues/385 to track the logging issues.

Reply:
Resolved via https://github.com/mdn/stumptown-content/pull/386